File: src/main/java/com/freya02/botcommands/api/core/service/annotations/ResolverFactory.java
package com.freya02.botcommands.api.core.service.annotations;

import com.freya02.botcommands.api.parameters.ParameterResolverFactory;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Marks this class as a {@link ParameterResolverFactory parameter resolver factory}.
 * <br>This is a specialization of {@link BService} for parameter resolver factories.
 *
 * <p><b>Requirement:</b> this must extend {@link ParameterResolverFactory}.
 *
 * @see ParameterResolverFactory
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.CLASS) //Read by ClassGraph
public @interface ResolverFactory { }

File: src/main/kotlin/com/freya02/botcommands/api/core/service/annotations/ResolverFactory.kt
package com.freya02.botcommands.api.core.service.annotations

import com.freya02.botcommands.api.parameters.ParameterResolverFactory

/**
 * Marks this class as a [parameter resolver factory][ParameterResolverFactory].<br>
 * This is a specialization of [BService] for parameter resolver factories.
 *
 * **Requirement:** this must extend [ParameterResolverFactory].
 *
 * **Warning:** Top-level functions are not processed, you must have them in an object/class.
 *
 * @see BService @BService
 *
 * @see ParameterResolverFactory
 */
@Target(AnnotationTarget.CLASS, AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER)
@Retention(AnnotationRetention.BINARY) //Read by ClassGraph
annotation class ResolverFactory  